File: bus/src/main/java/com/iwdael/livedatabus/LiveDataBus.java
package com.iwdael.livedatabus;

import android.annotation.SuppressLint;
import android.util.Log;

import androidx.arch.core.executor.ArchTaskExecutor;
import androidx.lifecycle.MutableBusLiveData;

import java.util.HashMap;
import java.util.Map;

public class LiveDataBus {
    private final Map<Object, Observable> busMap = new HashMap<>();

    private static class Holder {
        private static final LiveDataBus BUS = new LiveDataBus();
    }

    /**
     * 根据类型返回可观察的对象
     */
    public static <T> Observable<T> with(Class<T> type) {
        if (!Holder.BUS.busMap.containsKey(type)) {
            Holder.BUS.busMap.put(type, new ObservableWrapper<T>(new MutableBusLiveData<T>()));
        }
        return (Observable<T>) Holder.BUS.busMap.get(type);
    }

    public static <T> Observable<T> with(String type ) {
        if (!Holder.BUS.busMap.containsKey(type)) {
            Holder.BUS.busMap.put(type, new ObservableWrapper<T>(new MutableBusLiveData<T>()));
        }
        return (Observable<T>) Holder.BUS.busMap.get(type);
    }

    public static <T> void post(T e) {
        ObservableWrapper<T> observable = (ObservableWrapper<T>) with(e.getClass());
        observable.liveData.postValue(e);
    }

}

File: core/src/main/java/com/iwdael/livedatabus/LiveDataBus.kt
package com.iwdael.livedatabus

import android.os.Handler
import android.os.HandlerThread
import androidx.lifecycle.MutableBusLiveData
import java.util.HashMap

/**
 * author : Iwdael
 * e-mail : iwdael@outlook.com
 */
class LiveDataBus {
    private val busMap: MutableMap<Any, Observable<*>> = HashMap()
    private val handleThread = HandlerThread("live_data_bus")
    internal val handler: Handler

    private constructor() {
        handleThread.start()
        handler = Handler(handleThread.looper)
    }

    internal object Holder {
        val BUS = LiveDataBus()
    }

    companion object {
        @JvmStatic
        fun <V> with(type: Class<V>): Observable<V> {
            if (!Holder.BUS.busMap.containsKey(type)) {
                Holder.BUS.busMap[type] =
                    ObservableWrapper(MutableBusLiveData<V>())
            }
            return Holder.BUS.busMap[type] as Observable<V>
        }

        @JvmStatic
        fun <V> with(type: String): Observable<V> {
            if (!Holder.BUS.busMap.containsKey(type)) {
                Holder.BUS.busMap[type] = ObservableWrapper(MutableBusLiveData<V>())
            }
            return Holder.BUS.busMap[type] as Observable<V>
        }

        @JvmStatic
        fun <V : Any> post(value: V) {
            val observable = with(value.javaClass) as ObservableWrapper<V>
            observable.liveData.postValue(value)
        }

        @JvmStatic
        fun <V : Any> post(key: String, value: V) {
            val observable = with<V>(key) as ObservableWrapper<V>
            observable.liveData.postValue(value)
        }


    }
}